proposals/0055-csp.md

```diff
@@ -0,0 +1,251 @@
+
+- Start Date: 2025-05-13
+- Reference Issues: 
+- Implementation PR: 
+- Stage 2 Issue: https://github.com/withastro/roadmap/issues/1149
+- Stage 3 PR: https://github.com/withastro/roadmap/discussions/377
+
+# Summary
+
+Add a new feature that allows Astro assets, mainly scripts and styles, to support [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP).
+
+# Example
+
+```js
+// astro.config.mjs
+import { defineConfig } from "astro/config";
+
+export default defineConfig({
+    security: {
+        csp: true
+    }
+})
+```
+
+# Background & Motivation
+
+Include any useful background detail that that explains why this RFC is important.
+What are the problems that this RFC sets out to solve? Why now? Be brief!
+
+It can be useful to illustrate your RFC as a user problem in this section.
+(ex: "Users have reported that it is difficult to do X in Astro today.")
+
+# Goals
+
+- Support at least one solution (`nonce`, hash, etc.) that would support CSP.
+- Allow Astro scripts (client islands, server islands, view transitions, etc.) and Astro styles to work out of the box.
+- Update adapters to support the feature, if needed.
+- Make the feature opt-in, even when it's outside experimental.
+- Support the [dev server](#dev-server) (stretch)
+- View transitions (limited/opt-in support)
+
+# Non-Goals
+
+- Provide *multiple* solutions to fix the same problem
+- Support for third-party scripts dynamically imported by users e.g. `document.createElement("script")`, `import("https://path.to/script.js")`;
+- Support for third-party styles dynamically imported by users.
+- Performance. Users and authors are willing to overlook possible performance drawbacks in order to increase security.
+
+
+# Detailed Design
+
+## Solution adopted
+
+Contrary to other frameworks of the same ecosystem, our solution will rely on generating [**hashes of styles and scripts**](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP#hashes). For more information, please refer to the [alternatives](#alternatives) section.
+
+Astro provides the hashes during the rendering phase, they will be written in the `<head>` tag using the `<meta http-equiv="Content-Security-Policy">` meta tag.
+
+The *main* reason for using this strategy instead of `Response` headers is because this solution allows Astro to support two important use cases:
+- SSG, especially when no adapters are involved.
+- SPA (Single Page Applications), where scripts and styles *might* be loaded dynamically using client-side routing.
+
+## Generation of hashes
+
+The hashes will be generated using the [`crypto`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto) global object, and the package [`@oslojs/encoding`](https://encoding.oslojs.dev/). Astro *already uses* this package to encrypt props for server islands.
+
+Hashes will be generated using the `SHA-256` algorithm. Users can't change the algorithm.
```

Comment:
I would advise against hard coding SHA-256 as SHA-384 or SHA-512 are already deployed in the wild.
Following CSP spec would be a better option

https://developer.mozilla.org/en-US/docs/Glossary/Hash_function

Reply:
Are you suggesting that users provide their own function? Or are you suggesting something else?

Reply:
> Are you suggesting that users provide their own function? Or are you suggesting something else?

I'm suggesting that CSP integration should provide SHA-256, SHA-384 and SHA-512 as options for algorithm as they are pretty much standard across web dev space.

Also, I see now that `styleHashes` and `scriptHashes` sha256 requirement is being dropped, so I'm happy about it 👍